File: apps/web/src/features/accounts/utils/account-record-submission.test.ts
import { describe, expect, it } from "vitest"

import type { AccountRecordFormValues } from "../types/account-record"
import { resolveSubmissionAttempt } from "./refund-submission"
import { accountRecordSubmissionFingerprint } from "./account-record-submission"

const values: AccountRecordFormValues = {
  type: "income",
  accountId: "account",
  toAccountId: "",
  amount: "10.00",
  receivedAmount: "",
  mainCategoryId: "main",
  subcategoryId: "sub",
  occurredAt: "2026-09-23T10:30",
  notes: " note ",
}

describe("account record submission attempts", () => {
  it("reuses a key for the same normalized payload and rotates it when changed", () => {
    const first = resolveSubmissionAttempt(
      null,
      accountRecordSubmissionFingerprint(values),
      () => "key-1"
    )
    const unchanged = resolveSubmissionAttempt(
      first,
      accountRecordSubmissionFingerprint({
        ...values,
        amount: "10.0",
        notes: "note",
      }),
      () => "unexpected"
    )
    const changed = resolveSubmissionAttempt(
      unchanged,
      accountRecordSubmissionFingerprint({ ...values, amount: "11" }),
      () => "key-2"
    )

    expect(unchanged.idempotencyKey).toBe("key-1")
    expect(changed.idempotencyKey).toBe("key-2")
  })

  it("ignores fields that are not submitted for non-transfer records", () => {
    expect(
      accountRecordSubmissionFingerprint({
        ...values,
        toAccountId: "ignored",
        receivedAmount: "999",
      })
    ).toBe(accountRecordSubmissionFingerprint(values))
  })
})

File: apps/mobile/test/account_record_submission_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:tharwati_mobile/accounts/records/record_submission.dart';
import 'package:tharwati_mobile/accounts/records/records_models.dart';
import 'package:tharwati_mobile/core/idempotency_key.dart';

AccountRecordFormValues values({String amount = '10.00'}) =>
    AccountRecordFormValues(
      type: AccountRecordType.income,
      accountId: 'account',
      toAccountId: '',
      amount: amount,
      receivedAmount: '',
      mainCategoryId: 'main',
      subcategoryId: 'sub',
      occurredAt: '2026-09-23T10:30',
      notes: ' note ',
    );

void main() {
  test('unchanged normalized record payload reuses its idempotency key', () {
    final attempt = PayloadIdempotencyKey();
    final first = attempt.forPayload(
      accountRecordSubmissionFingerprint(values()),
    );
    final unchanged = attempt.forPayload(
      accountRecordSubmissionFingerprint(values(amount: '10.0')),
    );
    final changed = attempt.forPayload(
      accountRecordSubmissionFingerprint(values(amount: '11')),
    );

    expect(unchanged, first);
    expect(changed, isNot(first));
  });
}
